rfcs/0000-nixpkgs-breaking-change-policy.md

```diff
@@ -0,0 +1,75 @@
+
+# Summary
+
+This defines a policy for how to manage nixpkgs changes that break dependent derivations or tests. It specifies time frames and procedures so that maintainers of dependencies and dependents can know what to expect from each other when making breaking changes.
+
+This document is only focused on changes to derivation builds and tests it **does not** aim to make any opinion on changes that can break user machines or configurations.
+
+# Motivation
+
+Right now the procedure for making breaking changes varies widely by maintainer and is largely tribal knowledge. There are a wide variety of approaches ranging from:
+
+- After testing to confirm that a change is correct, merge to `master` and notify dependents that they need to update their packages.
+- Merge to `staging`, notify dependents then eventually merge to `master`.
+- Gather a large number of fixes in the PR branch then merge to `master` once most or all dependencies are fixed.
+
+This aims to provide a uniform approach so that everyone involved knows what to expect and to allow further workflows including automation to be build on top of these procedures.
+
+## Past Discussions
+
+https://discourse.nixos.org/t/please-fix-all-packages-which-were-broken-by-the-qt-wrapping-changes/6444
+
+# Detailed design
+
+## Goals
+
+The following procedure will be used for making breaking changes. It has a few primary goals:
+
+- Avoid putting more burden than necessary on the dependency maintainer. If the maintainers of core derivations face toil proportionally to the number of transitive dependencies they will quickly become overloaded. These maintainers are arguably the most critical to nixpkgs and  their load needs to be kept manageable.
+- Avoid unnecessarily breaking packages for any period of time. There are a number of users on the `*-unstable` channels and it is annoying if packages in their configuration get broken.
+- Avoid breaking the `master` branch as much as possible. The `master` branch is used by the `*-unstable` channels and breaking this branch means that no updates can be pushed. This includes critical security updates as well as regular feature updates.
+
+## Procedure
+
+1. The maintainer will prepare a PR with their intended changes.
+2. The maintainer should test a sample of dependent derivations to ensure that their change is not unnecessarily or unintentionally breaking. (Example: Ensure that a critical output file was not forgotten) Note that sometimes it **is** necessary to break all dependent packages, and the maintainer is not required to avoid this.
+3. The maintainer will get the PR reviewed and approved.
+    1. It is **recommended but not required** to have some maintainers of dependent packages involved in the review to review if the breakage is justified.
+4. The maintainer will contact the maintainers of all dependent, broken packages (herein called sub-maintainers).
```

Comment:
How? Should I collect email addresses and write them an email? I think this is not practically possible.

Reply:
The idea was to tag them on the PR. We have contact info in https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix so this can (and is expected to be) automated.

Reply:
I know this pops up quite often recently (or maybe I'm wrong and it's just selective perception), but this is something that can be solved by Hydra notifications or e.g. https://github.com/NixOS/hydra/pull/743.

Also I'd like to emphasize that there's a big difference between the stability of `master` and release branches, I mean there's a reason why `nixos-unstable` has `unstable` in its name and we have ZHF iterations after branch-off.

Reply:
GitHub also limits the amount of mentions in case many people are affected.

Reply:
Part of the idea here is to allows us to move towards keeping the unstable branches green (at least from the building point of view). This allows security fixes and other changes to get pushed at any time. This policy won't get us all of the way there but it already in theory does as all broken packages should be marked as broken before getting merged to `master`.

Hydra notifications would be good, but they only apply to the merged branch. At that point it is too late, that broken package is already blocking changes which is bad for critical fixes and annoying to contributors (as their change gets delayed). Ideally this would be the same sort of tech, but applied somewhat pre-merge.

I don't know if there is an official stance on what `-unstable` means. Because the channel is always tested on hydra before being published. My understanding it is more that packages, including major versions will be updated frequently. I don't think it should mean "fails to build and release for days or weeks at a time".

Reply:
Well, the channel doesn't break on **any** package though. Most of the leaf-packages (and AFAIU that's what we're talking about here) can be totally broken (i.e. failing to compile or whatever) without preventing `nixos-unstable` from moving forward.

Reply:
That's a good point. For what I am aiming for I don't really care about a package if the channel config doesn't care about it. On the other hand is there a downside to marking these broken packages as broken? If there is it would be easy enough to update the RFC to define what dependents we care about and just ignore them completely (maybe still notify them, but no need to mark as broken).

Reply:
> We have contact info in

Those information are optional and not all people added a github name or an email address.

Reply:
Well if they aren't contactable than their package will be broken. I don't think we can really do anything about this if we don't have any handles to the maintainer.

Reply:
We should add something similar to ``if you want to be contacted you need to add contact information because otherwise it is not possible``.

Reply:
How does this sound? https://github.com/NixOS/rfcs/pull/88/files#r602721463